### PR TITLE
chore(test): Migrate test suites from mocha to node test runner

### LINKED ISSUE #157: Switch to node --test from mocha.
Mocha is a pain to configure for ESM and Typescript. Since we're switcing to ESM, now is a good time to swap out Mocha and switch to `node --test`, which makes this a breeze.

### PR DESCRIPTION
**Pull request by Zen-cronic**
Resolves #157 & #158 

### Changes

- Replaced `mocha` with `node --test` script 
- Replaces `expect` from `chai` with `node:assert`

### CODE REVIEW

**Comment by Zen-cronic** [comment]
The CI tests are ran on node-version: [14.x, 16.x, 18.x, 20.x, 22.x]. But node test runner was added in [v16.17.0](https://nodejs.org/api/test.html#test-runner). So would support for node 14 be dropped?

**Comment by evert** [comment]
Nice! This is awesome. I think it's fine to finally drop Node 14. It's quite old at this point.

**Comment by Zen-cronic** [comment]
Interestingly, the tests are not run or only run partially, depending on the node version.

- `v16.20.2`: no tests run and hanging.
- `v18.20.4`: only the first (alphabetically) test file is run `authorization-code.ts` and left hanging.
- `v20.17.0`: same as v18
- `v22.8.0`: same as v18

*Tested locally and as observed in CI

**Comment by evert** [comment]
Good catch! I recently migrated a different library to `npm --test` and this is how I've solved that:

https://github.com/badgateway/structured-headers/pull/58/files

Basically I:

* Installed `tsx` (and ditched `ts-node`, which doesn't work well with ESM anyway).
* Instead of running `node --test`, run `tsx --test test/**/*.ts`

This mostly solved all my problems. `tsx` is effectively a wrapper around `node` that sets up parsing Typescript transparently.

**Comment by Zen-cronic** [comment]
Oh I see. Will try it!

**Comment by Zen-cronic** [comment]
> Interestingly, the tests are not run or only run partially, depending on the node version.
> 

The tests were passing but the process was hanging. It seems that `node:test` isn't as dev-ready as other mature test frameworks. All connections/promises had to be closed explicitly like [this](https://github.com/badgateway/oauth2-client/blob/385de71c17d625b3f416a831a88f5c3716cec136/test/client-credentials.ts#L7-L12). Whereas with mocha, the process ended as expected after the tests pass/fail. 

Currently, the tests in [fetch-wrapper.ts](https://github.com/badgateway/oauth2-client/blob/385de71c17d625b3f416a831a88f5c3716cec136/test/fetch-wrapper.ts) will all pass but the process will hang, so the next test file isn't continued. More specifically, the first two tests in that file are causing the hang - verified by attaching `only` and running locally. Will investigate further.

File: package.json [hunk 1]
```diff
@@ -37,17 +37,13 @@
...
     "@types/node": "^18.0.0",
     "@typescript-eslint/eslint-plugin": "^6.2.0",
     "@typescript-eslint/parser": "^6.2.0",
-    "chai": "^4.3.6",
     "eslint": "^8.1.0",
-    "mocha": "^10.0.0",
     "node-fetch": "^2.6.7",
     "ts-loader": "^9.2.6",
-    "ts-node": "^10.7.0",
+    "tsx": "^3.14.0",
```
Comment (Zen-cronic) [inline package.json:46]: Using v3 b/c node-16 support dropped in semver [v4](https://github.com/privatenumber/tsx/releases/tag/v4.0.0).

**Comment by evert** [comment]
> The tests were passing but the process was hanging. It seems that node:test isn't as dev-ready as other mature test frameworks. All connections/promises had to be closed explicitly like [this](https://github.com/badgateway/oauth2-client/blob/385de71c17d625b3f416a831a88f5c3716cec136/test/client-credentials.ts#L7-L12). Whereas with mocha, the process ended as expected after the tests pass/fail.

I'd imagine that this is a deliberate choice by Node, to keep things somewhat simple and explicit, but yeah i bet it was annoying to deal with.

> Currently, the tests in [fetch-wrapper.ts](https://github.com/badgateway/oauth2-client/blob/385de71c17d625b3f416a831a88f5c3716cec136/test/fetch-wrapper.ts) will all pass but the process will hang, so the next test file isn't continued. More specifically, the first two tests in that file are causing the hang - verified by attaching only and running locally. Will investigate further.

I'm almost certain this is because `fetch-wrapper` calls `setTimeout` which keeps an event open.

One idea is to explicitly disable this, or call `clearTimeout` in the cleanup function for this. (you may need to cast that object to `any` first to ignore the fact that it's a `private` function).

There's definitely more hoops here than I expected, so I understand if you're out of patience. If so I'll definitely build on your work to complete this. Either way I really appreciate the work so far!

**Comment by Zen-cronic** [comment]
> 
> I'd imagine that this is a deliberate choice by Node, to keep things somewhat simple and explicit, but yeah i bet it was annoying to deal with.
> 
> 
> I'm almost certain this is because `fetch-wrapper` calls `setTimeout` which keeps an event open.
> 
> One idea is to explicitly disable this, or call `clearTimeout` in the cleanup function for this. (you may need to cast that object to `any` first to ignore the fact that it's a `private` function).
> 
Gotcha, I added a fix. Feel free to adjust for any improvement.

**Comment by evert** [comment]
Very frustrating! If I'm reading the errors right, we need tsx 3 for Node =< 16, and tsx 4 for Node > 20

**Comment by evert** [comment]
Given that dropping older Node versions is already a major change, maybe it's time to also drop Node 16, and make this PR of a future 3.0 release.

**Pull request merged**